### ISSUE TITLE
maxp.maxSizeOfInstructions is not calculated?

### ISSUE THREAD

**Issue by khaledhosny** [fonttools/fonttools#902]
I was trying to make a font with defective `glyf` table glyphs for testing OTS, when I found that `Tests/ttx/data/TestTTF.ttf` is already rejected by OTS. This seems to be because `maxSizeOfInstructions` is set to 0, but some glyphs have `instructionLength` of 2.

It seems `maxSizeOfInstructions` is not calculated when generating `maxp` table, I’m not sure if this is intentional or not.

**Comment by anthrotype** [comment]
Yes, we currently don't recalculate that. We probably should.

The problem is that there is some ambiguity as to how properly calculate that value.
I myself raised the issue with the OpenType mailing list almost two years ago, and Greg Hitchcock from Microsoft replied with a long and detailed history of this maxp field.

TLDR...

This should be based on the largest set of instructions (in the glyf table) for a single simple or composite glyph.

Tools like VTT or ttfautohint also consider `fpgm` and `prep` when computing this value, but apparently are wrong.

FontValidator also warns when the value does not match the "Maximum byte count for glyph instructions" (as the `maxp` spec puts it).

**Comment by lemzwerg** [comment]
OK, just rereading Greg's e-mail...

The sentence 'Tools like VTT or ttfautohint [...] are apparently wrong' is highly debatable :-) Given that no implementation is actually using the max value, and that the (former) de-facto tool for creating bytecode (VTT)  disagrees or disagreed for a long time with Greg's interpretation, and that the wording in the specification is still ambiguous in the newest OpenType specification, we have to live with the dichotomy of two (or even more) valid values.  Using the maximum value of the bytecode sizes in the `fpgm`, `prep`, and `glyf` does certainly work.  Admittedly, it might be too large, but it doesn't do any harm.

**Comment by behdad** [comment]
> I was trying to make a font with defective glyf table glyphs for testing OTS, when I found that Tests/ttx/data/TestTTF.ttf is already rejected by OTS. This seems to be because maxSizeOfInstructions is set to 0, but some glyphs have instructionLength of 2.

Relax ots I'd say.

> It seems maxSizeOfInstructions is not calculated when generating maxp table, I’m not sure if this is intentional or not.

Maybe sometimes it should.  I don't want it to happen in subetter though.  Maybe same as other recalc stuff?  Also, it will force expansion of all glyf entries which is costly.  We can implement a fastpath that doesn't expand.  I did that for some other stuff in the subsetter.

**Comment by madig** [comment]
How bout I send a PR where we use Greg's definition for the value? If any strategy is fine, we just take the simplest one.

**Comment by madig** [comment]
Closing, as

1) There are multiple ways of calculating it and it's not clear which way fontTools should go
2) There are no known users of the value
3) Users and compilers can set the value to whatever they need
